File: api/src/main/kotlin/cn/enaium/bookstore/controller/VoteController.kt
/*
 * Copyright (c) 2025 Enaium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package cn.enaium.bookstore.controller

import cn.enaium.bookstore.error.VoteException
import cn.enaium.bookstore.model.entity.dto.VoteInput
import cn.enaium.bookstore.model.entity.dto.VoteView
import cn.enaium.bookstore.service.VoteService
import org.babyfish.jimmer.Page
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.util.*

/**
 * @author Enaium
 */
@RestController
@RequestMapping("/votes")
class VoteController(
    val voteService: VoteService
) {
    @PreAuthorize("isAuthenticated()")
    @GetMapping
    fun getVotes(
        @RequestParam(value = "index", defaultValue = "0") index: Int = 0,
        @RequestParam(value = "size", defaultValue = "10") size: Int = 10,
        @RequestParam(value = "type") type: VoteType
    ): Page<VoteView> {
        return voteService.getVotes(index, size, type)
    }

    @GetMapping("/state")
    fun state(
        @RequestParam(value = "issuerId", required = false) issuerId: UUID? = null,
        @RequestParam(value = "bookId", required = false) bookId: UUID? = null,
        @RequestParam(value = "authorId", required = false) authorId: UUID? = null,
        @RequestParam(value = "commentId", required = false) commentId: UUID? = null
    ): VoteView? {
        return voteService.state(issuerId, bookId, authorId, commentId)
    }

    @PreAuthorize("hasAnyRole('ROLE_MODERATOR', 'ROLE_USER')")
    @PutMapping
    fun save(@RequestBody input: VoteInput) {
        voteService.save(input)
    }

    @PreAuthorize("hasAnyRole('ROLE_MODERATOR', 'ROLE_USER')")
    @DeleteMapping("/{id}")
    @Throws(VoteException::class)
    fun delete(@PathVariable id: UUID) {
        voteService.delete(id)
    }

    enum class VoteType {
        ISSUER,
        BOOK,
        AUTHOR,
        COMMENT,
    }
} 

File: web/src/__generated/services/VoteController.ts
import type {Executor} from '../';
import type {VoteType} from '../model/enums/';
import type {Page, VoteInput, VoteView} from '../model/static/';

export class VoteController {
    
    constructor(private executor: Executor) {}
    
    readonly delete: (options: VoteControllerOptions['delete']) => Promise<
        void
    > = async(options) => {
        let _uri = '/votes/';
        _uri += encodeURIComponent(options.id);
        return (await this.executor({uri: _uri, method: 'DELETE'})) as Promise<void>;
    }
    
    readonly getVotes: (options: VoteControllerOptions['getVotes']) => Promise<
        Page<VoteView>
    > = async(options) => {
        let _uri = '/votes';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.index;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'index='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.type;
        _uri += _separator
        _uri += 'type='
        _uri += encodeURIComponent(_value);
        _separator = '&';
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<Page<VoteView>>;
    }
    
    readonly save: (options: VoteControllerOptions['save']) => Promise<
        void
    > = async(options) => {
        let _uri = '/votes';
        return (await this.executor({uri: _uri, method: 'PUT', body: options.body})) as Promise<void>;
    }
    
    readonly state: (options: VoteControllerOptions['state']) => Promise<
        VoteView | undefined
    > = async(options) => {
        let _uri = '/votes/state';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.issuerId;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'issuerId='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.bookId;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'bookId='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.authorId;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'authorId='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.commentId;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'commentId='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<VoteView | undefined>;
    }
}

export type VoteControllerOptions = {
    'getVotes': {
        index?: number | undefined, 
        size?: number | undefined, 
        type: VoteType
    }, 
    'state': {
        issuerId?: string | undefined, 
        bookId?: string | undefined, 
        authorId?: string | undefined, 
        commentId?: string | undefined
    }, 
    'save': {
        body: VoteInput
    }, 
    'delete': {
        id: string
    }
}
